vmware-tanzu/load-balancer-operator-for-kubernetes
# Fix akodeploymentconfig cluster selector didn't select the right clusters issue

**What this PR does / why we need it**:
- Fix `akodeploymentconfig.spec.clusterSelector` can't select the right clusters issue, with this fix, load-balancer-operator will follow the following rules to find matched workload clusters:
    - AKODeploymentConfig object will normally select all the clusters matched rules defined in `spec.clusterSelector`, the following are **exceptions**:
    - the default AKODeploymentConfig object `install-ako-for-all` with an empty selector can select all the workload clusters
    - The specific AKODeploymentConfig `install-ako-for-management-cluster` can only select management cluster.
    - When there are two AKODeploymentConfig objects selecting the same cluster, it follows the below rules:
        - Default `install-ako-for-all` with empty selector has the lowest priority on this cluster, and it can be overridden by other akodeploymentconfig objects.
        - New created AKODeploymentConfig object will skip those clusters which already been selected by other AKODeploymentConfig objects.

- Remove `networking.tkg.tanzu.vmware.com/avi-skip-default-adc` label
- Update cluster's `networking.tkg.tanzu.vmware.com/avi` label, it will be filled with the akodeployment object name which currently selects this cluster.
- Remove redundant deleting add-on secret code, add-on secret will be managed by tanzu-add-on-manager, the delete add-on secret code here never work.

**Which issue(s) this PR fixes**:

Fixes #

**Describe testing done for PR**:


- [x] unit test cases passed.
- [x] integration test cases passed. 
- [x] Manually e2e tested following test cases on a vSphere testbed:
    - Default adc with empty selector:
         - Should select all workload clusters without custom adc matched. 
         - Create a custom adc with selector -> custom adc selects all should be selected, others should be selected by default adc
         - Create two custom adc with selector, no conflict -> custom adc selects all workload clusters with matched labels, others should be selected by default adc
         - update cluster label without conflict-> cluster switch adc to new matched one as expected.
         - set `AVI_LABELS` when creating cluster -> should be selected by expected custom adc
         - update cluster label got selected by two adc -> can only override default adc with empty selector
         - cluster already has avi label and only be selected by default adc, then create an adc matching cluster label -> cluster should update to be selected by new adc.
         - cluster already has avi label but already be selected by another adc, then create an adc matching cluster label  -> can only override default adc with empty selector, otherwise, it should not override the previous one
    - Default adc with selector:
       - should select expected cluster, not select didn’t match clusters
       - create a custom adc with selector, no conflict -> each should select matched clusters
       -  update cluster label without conflict -> should be selected by new matched one
       - set `AVI_LABELS` when creating cluster -> should be selected by expected custom adc
       - create new adc with selector, should select matched and previously non-selected cluster
       - cluster already has avi label and only be selected by default adc, then create an adc matching cluster label -> should not override since default adc didn't select all in this case. 
       - cluster already has avi label but already be selected by another adc, then create an adc matching cluster label  -> should not override. 

**Special notes for your reviewer**:

**Release note**:

```release-note
removed networking.tkg.tanzu.vmware.com/avi-skip-default-adc label and updated networking.tkg.tanzu.vmware.com/avi label with akodeploymentconfig name
```
**New PR Checklist**

- [x] Add appropriate [labels](https://github.com/vmware-samples/load-balancer-operator-for-kubernetes/labels) according to what type of issue is being addressed.
